hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java

```diff
@@ -937,141 +946,111 @@ public String startCommit() {
-   * asynchronously. Please always use this serially before or after an insert/upsert action.
+   * Schedules a new compaction instant with passed-in instant time
+   * @param instantTime     Compaction Instant Time
+   * @param extraMetadata   Extra Metadata to be stored
    */
-  public JavaRDD<WriteStatus> compact(String commitTime) throws IOException {
-    // Create a Hoodie table which encapsulated the commits and files visible
-    HoodieTable<T> table = HoodieTable.getHoodieTable(
-        new HoodieTableMetaClient(jsc.hadoopConfiguration(), config.getBasePath(), true), config);
-    // TODO : Fix table.getActionType for MOR table type to return different actions based on delta or compaction
-    writeContext = metrics.getCommitCtx();
-    JavaRDD<WriteStatus> statuses = table.compact(jsc, commitTime);
-    // Trigger the insert and collect statuses
-    statuses = statuses.persist(config.getWriteStatusStorageLevel());
-    String actionType = HoodieActiveTimeline.COMMIT_ACTION;
-    commitOnAutoCommit(commitTime, statuses, actionType);
-    return statuses;
+  public void scheduleCompactionWithTime(String instantTime, Optional<Map<String, String>> extraMetadata)
+      throws IOException {
+    HoodieTableMetaClient metaClient = new HoodieTableMetaClient(jsc.hadoopConfiguration(),
+        config.getBasePath(), true);
+    // if there are inflight ingestions, their instantTime must not be less than that of compaction instant time
+    metaClient.getActiveTimeline().filterInflightsExcludingCompaction().firstInstant().ifPresent(earliestInflight -> {
+      Preconditions.checkArgument(
+          HoodieTimeline.compareTimestamps(earliestInflight.getTimestamp(), instantTime, HoodieTimeline.GREATER),
+          "Earliest ingestion inflight instant time must be later "
+              + "than compaction time. Earliest :" + earliestInflight + ", Compaction scheduled at " + instantTime);
+    });
+    HoodieTable<T> table = HoodieTable.getHoodieTable(metaClient, config);
+    HoodieCompactionWorkload workload = table.scheduleCompaction(jsc, instantTime);
+    extraMetadata.ifPresent(workload::setExtraMetadata);
+    HoodieInstant compactionInstant =
+        new HoodieInstant(State.REQUESTED, HoodieTimeline.COMPACTION_ACTION, instantTime);
+    metaClient.getActiveTimeline().saveToCompactionRequested(compactionInstant,
+        AvroUtils.serializeCompactionWorkload(workload));
   }
 
   /**
-   * Commit a compaction operation
+   * Return all pending compactions with instant time for clients to decide what to compact next.
+   * @return
    */
-  public void commitCompaction(String commitTime, JavaRDD<WriteStatus> writeStatuses,
-      Optional<HashMap<String, String>> extraMetadata) {
-    String commitCompactionActionType = HoodieActiveTimeline.COMMIT_ACTION;
-    commit(commitTime, writeStatuses, extraMetadata, commitCompactionActionType);
+  public List<HoodieCompactionInstantWithWorkload> getPendingCompactions() {
```

Comment:
Should this be moved to the HoodieReadClient ?

Reply:
Done. Moved to ReadClient